File: Zds/Zds.Flow/Updaters/UpdaterX.vb
Imports Zds.Flow.Collections
Imports Zds.Flow.Updatables

Namespace Updaters
    <DebuggerStepThrough>
    Public Class UpdaterX
        Inherits Updater
        Private _ThreadCreatedEvent As SafeList(Of UpdaterXEventDelegate)
        Private _ThreadDestroyedEvent As SafeList(Of UpdaterXEventDelegate)
        Private _Threads As New SafeList(Of Threading.Thread)
        Private _TotalThreads As Integer = 0
        Private _IdleThreads As Integer = 0
        Private _AvailableThreads As Integer = 0
        Public Custom Event ThreadCreatedEvent As UpdaterXEventDelegate
            AddHandler(value As UpdaterXEventDelegate)
                If _ThreadCreatedEvent Is Nothing Then _ThreadCreatedEvent = New SafeList(Of UpdaterXEventDelegate)
                _ThreadCreatedEvent.Add(value)
            End AddHandler
            RemoveHandler(value As UpdaterXEventDelegate)
                If _ThreadCreatedEvent IsNot Nothing Then _ThreadCreatedEvent.Remove(value)
            End RemoveHandler
            RaiseEvent(Sender As UpdaterX, Thread As Threading.Thread)
                If _ThreadCreatedEvent IsNot Nothing Then
                    For Each [Delegate] In _ThreadCreatedEvent.Elements
                        [Delegate].Invoke(Sender, Thread)
                    Next
                End If
            End RaiseEvent
        End Event
        Public Custom Event ThreadDestroyedEvent As UpdaterXEventDelegate
            AddHandler(value As UpdaterXEventDelegate)
                If _ThreadDestroyedEvent Is Nothing Then _ThreadDestroyedEvent = New SafeList(Of UpdaterXEventDelegate)
                _ThreadDestroyedEvent.Add(value)
            End AddHandler
            RemoveHandler(value As UpdaterXEventDelegate)
                If _ThreadDestroyedEvent IsNot Nothing Then _ThreadDestroyedEvent.Remove(value)
            End RemoveHandler
            RaiseEvent(Sender As UpdaterX, Thread As Threading.Thread)
                If _ThreadDestroyedEvent IsNot Nothing Then
                    For Each [Delegate] In _ThreadDestroyedEvent.Elements
                        [Delegate].Invoke(Sender, Thread)
                    Next
                End If
            End RaiseEvent
        End Event
        Public Property MaxThreads As Integer = 50
        Public Property MaxIdleThreads As Integer = 10
        Public Property MinThreadLifeSpan As TimeSpan = TimeSpan.FromMinutes(1)
        Public Property MaxThreadLifeSpan As TimeSpan = TimeSpan.MaxValue
        Protected Property ActionLock As New Object
        Public ReadOnly Property Threads As Threading.Thread()
            Get
                Return _Threads.Elements
            End Get
        End Property
        Public ReadOnly Property TotalThreads As ULong
            Get
                Return _TotalThreads
            End Get
        End Property
        Public ReadOnly Property IdleThreads As ULong
            Get
                Return _IdleThreads
            End Get
        End Property
        Public Overrides Sub Start()
            SyncLock ActionLock
                If IsDestroyed Then Exit Sub
                If Not IsRunning Then
                    MyBase.Start()
                    If Targets.Length <> 0 And TotalThreads = 0 Then CreateThread()
                End If
            End SyncLock
        End Sub
        Public Overrides Sub [Stop]()
            SyncLock ActionLock
                If IsDestroyed Then Exit Sub
                If IsRunning Then
                    MyBase.Stop()
                End If
            End SyncLock
        End Sub
        Public Overrides Sub Destroy()
            SyncLock ActionLock
                MyBase.Destroy()
            End SyncLock
        End Sub
        Public Sub Serve()
            If IsDestroyed Then Exit Sub
            If Not IsRunning Then Exit Sub
            Work()
        End Sub
        Public Sub Serve(state As Object)
            Work()
        End Sub
        Public Sub CreateThread()
            If IsDestroyed Then Exit Sub
            If Not IsRunning Then Exit Sub
            Dim Thread As New Threading.Thread(AddressOf Work)
            Thread.Start()
        End Sub
        Public Overrides Sub Add(Updatable As IUpdatable)
            MyBase.Add(Updatable)
            If Targets.Length <> 0 And TotalThreads = 0 Then CreateThread()
        End Sub
        Public Overrides Sub Remove(Updatable As IUpdatable)
            MyBase.Remove(Updatable)
        End Sub
        Protected Overridable Sub Work()
            If IsDestroyed Then Exit Sub
            If Not IsRunning Or Targets.Length = 0 Then Exit Sub
            Dim Thread As Threading.Thread = Threading.Thread.CurrentThread
            RaiseEvent ThreadCreatedEvent(Me, Thread)
            _Threads.Add(Thread)
            Threading.Interlocked.Increment(_TotalThreads)
            Threading.Interlocked.Increment(_IdleThreads)
            Dim Stopwatch As New Stopwatch.Stopwatch
            Stopwatch.Start()
            Dim IdleThreads As Integer
            Do
                If Not IsRunning Or Targets.Length = 0 Then Exit Do
                If IsPaused = False Then
                    IdleThreads = Threading.Interlocked.Decrement(_IdleThreads)
                    If IdleThreads = 0 AndAlso _TotalThreads <= _MaxThreads Then
                        CreateThread()
                    End If
                    Update()
                    IdleThreads = Threading.Interlocked.Increment(_IdleThreads)
                    If (IdleThreads > _MaxIdleThreads AndAlso Stopwatch.Elapsed >= MinThreadLifeSpan) OrElse
                        (_TotalThreads > 1 AndAlso Stopwatch.Elapsed >= MaxThreadLifeSpan) Then
                        Exit Do
                    End If
                End If
                If Not IsRunning Or Targets.Length = 0 Then Exit Do
                If DelayHandler IsNot Nothing Then DelayHandler.Delay()
            Loop
            Stopwatch.Stop()
            _Threads.Remove(Thread)
            Threading.Interlocked.Decrement(_IdleThreads)
            Threading.Interlocked.Decrement(_TotalThreads)
            RaiseEvent ThreadDestroyedEvent(Me, Thread)
        End Sub

        Public Delegate Sub UpdaterXEventDelegate(Sender As UpdaterX, Thread As Threading.Thread)
    End Class
End Namespace

File: Flow/Flow/Updaters/UpdaterX.cs
using Flow.Collections;
using Flow.Updatables;
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;

namespace Flow.Updaters
{
    public class UpdaterX : Updater
    {
        private SafeList<Thread> threads = new SafeList<Thread>();
        private int totalThreads = 0;
        private int idleThreads = 0;
        public override event ThreadEventDelegate ThreadCreated;
        public override event ThreadEventDelegate ThreadDestroyed;
        public int TotalThreads { get => totalThreads; }
        public int IdleThreads { get => idleThreads; }
        public int SleepTime { get; set; } = 1;
        public int MaxThreads { get; set; } = 50;
        public int MaxIdleThreads { get; set; } = 10;
        public TimeSpan MinThreadLifeSpan { get; set; } = TimeSpan.FromSeconds(10);
        
        public void CreateThread()
        {
            if (!IsRunning || Targets.Length == 0) return;
            var thread = new Thread(Work);
            thread.Start();
        }
        private void Work()
        {
            if (!IsRunning || Targets.Length == 0) return;
            var thread = Thread.CurrentThread;
            threads.Add(thread);
            var _totalThreads = Interlocked.Increment(ref totalThreads);
            if (_totalThreads > MaxThreads)
            {
                Interlocked.Decrement(ref totalThreads);
                return;
            }
            Interlocked.Increment(ref idleThreads);
            ThreadCreated?.Invoke(this, thread);
            var Stopwatch = new Stopwatch.Stopwatch();
            Stopwatch.Start();
            int _idleThreads = 0;
            while (IsRunning && Targets.Length > 0)
            {
                _idleThreads = Interlocked.Decrement(ref idleThreads);
                if (idleThreads == 0 && totalThreads < MaxThreads) CreateThread();
                Update();
                _idleThreads = Interlocked.Increment(ref idleThreads);
                if (_idleThreads >= MaxIdleThreads && Stopwatch.Elasped >= MinThreadLifeSpan) break;
                if (!IsRunning || Targets.Length == 0) break;
                Delay();
            }
            Stopwatch.Stop();
            Interlocked.Decrement(ref idleThreads);
            Interlocked.Decrement(ref totalThreads);
            threads.Remove(thread);
            ThreadDestroyed?.Invoke(this, thread);
        }
        protected virtual void Delay() => Thread.Sleep(SleepTime);

        public override void Add(IUpdatable obj)
        {
            base.Add(obj);
            if (IsRunning && totalThreads == 0) CreateThread();
        }
        public override void Start()
        {
            base.Start();
            if (totalThreads == 0) CreateThread();
        }
        public override void Reset()
        {
            Stop();
            Wait();
        }

        public override bool Wait(TimeSpan timeout)
        {
            Stopwatch.Stopwatch stopwatch = new Stopwatch.Stopwatch();
            stopwatch.Start();
            while (totalThreads != 0)
            {
                Thread.Sleep(1);
                if (totalThreads == 0) break;
                if (stopwatch.Elasped >= timeout)
                {
                    stopwatch.Stop();
                    return false;
                }
            }
            stopwatch.Stop();
            return true;
        }

        public delegate void ThreadEvent(object sender, Thread thread);
    }
}
